paradedb/paradedb
# feat: `pg_dump` support + schema evolution

# Ticket(s) Closed

- Closes #

## What
Implements the tableAM scan methods, which are used by the `COPY to STDOUT` command (invoked by pg_dump). Also handles `ADD/DROP/RENAME` column correctly.

How to test:

1. Inside `cargo pgrx run`, create a deltalake table.

```sql
CREATE TABLE t (a int, b text) USING deltalake;
INSERT INTO t VALUES (1, 'test'), (2, 'another');
```

2. Exit the connection and run a `pg_dump`

```bash
pg_dump -U <username> -d pg_analytics -h localhost -p 28816 > sample_db_backup.sql
```

Make sure the replace `username` with your username. This will create a dump of your database to `sample_db_backup.sql`.

3. Reconnect to the database and drop the table you just created.

```sql
DROP TABLE t;
```

4. Exit the connection and run the backup script

```bash
psql -U<username> -d pg_analytics -h localhost -p 28816 -f sample_db_backup.sql
```

5. Reconnect to the database to see the dropped table restored.

```sql
SELECT * FROM t;
```

## Why
Enables users to back up deltalake tables with `pg_dump`.

## How
Two changes:

1. Implemented the table AM scan functions, which are used by pg_dump to read from the table.
2. Fixed the way we were handling `ALTER` statements on deltalake tables to allow more statements to pass through the Postgres. Also enabled users to run `ADD COLUMN` over deltalake tables.

## Tests
See new tests.
